Allow time for propagation of new IAM resources on KMS policy update

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform CLI and Terraform AWS Provider Version

Terraform v1.1.5
Provider v4.14.0 (HEAD: cffbd38)

### Affected Resource(s)

* `aws_kms_key`

### Terraform Configuration Files

Given the distributed nature of AWS, IAM changes take time to propagate. During the creation of KMS keys and their resource policies, a `MalformedPolicyDocumentException` indicates that IAM changes might not have propagated, and this is handled by the provider, which retries for up to two minutes.

However, when updating the resource policy of an already existing key, the provider fails to handle the same error.

```hcl
data "aws_caller_identity" "this" {}

#resource "aws_iam_role" "example_role" {
#  name               = "example_role"
#  assume_role_policy = data.aws_iam_policy_document.assume_example_role.json
#}

#data "aws_iam_policy_document" "assume_example_role" {
#  statement {
#    actions = ["sts:AssumeRole"]
#    principals {
#      type        = "AWS"
#      identifiers = [data.aws_caller_identity.this.arn]
#    }
#  }
#}

resource "aws_kms_key" "example_key" {
  policy      = data.aws_iam_policy_document.example_key_policy.json
}

data "aws_iam_policy_document" "example_key_policy" {
#  statement {
#    sid    = "Decrypt"
#    effect = "Allow"
#    principals {
#      type        = "AWS"
#      identifiers = [aws_iam_role.example_role.arn]
#    }
#    actions = [
#      "kms:Decrypt",
#    ]
#    resources = ["*"]
#  }

  statement {
    sid    = "Administrate"
    effect = "Allow"
    principals {
      type = "AWS"
      identifiers = [data.aws_caller_identity.this.arn]
    }
    actions = [
      "kms:*"
    ]
    resources = ["*"]
  }
}
```

### Debug Output

```
2022-05-08T18:53:33.578+0100 [ERROR] vertex "aws_kms_key.example_key" error: error updating KMS Key (xxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx) policy: MalformedPolicyDocumentException: Policy contains a statement with one or more invalid principals.
|
│ Error: error updating KMS Key (xxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx) policy: MalformedPolicyDocumentException: Policy contains a statement with one or more invalid principals.
│ 
│   with aws_kms_key.example_key,
│   on kms_update_wait_iam.tf line 18, in resource "aws_kms_key" "example_key":
│   18: resource "aws_kms_key" "example_key" {
│ 
```

### Expected Behavior

As with key creation, the provider should retry the key update until the timeout allowed for IAM propagation (two minutes) is reached. 

### Actual Behavior

The provider doesn't retry, instead failing on the first `MalformedPolicyDocumentException`.

### Steps to Reproduce

1. `terraform apply` the Terraform configuration from above to create a KMS key
2. Uncomment the commented lines and `terraform apply` again to create a new IAM role and reference it in the policy

### References

* #7646

--- a/.changelog/24697.txt
+++ b/.changelog/24697.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+resource/aws_kms_key: Retry on `MalformedPolicyDocumentException` errors when updating key policy
+```

--- a/internal/service/kms/key.go
+++ b/internal/service/kms/key.go
@@ -444,7 +444,7 @@ func updateKeyPolicy(conn *kms.KMS, keyID string, policy string, bypassPolicyLoc
 		return nil, err
 	}
 
-	_, err = tfresource.RetryWhenAWSErrCodeEquals(PropagationTimeout, updateFunc, kms.ErrCodeNotFoundException)
+	_, err = tfresource.RetryWhenAWSErrCodeEquals(PropagationTimeout, updateFunc, kms.ErrCodeNotFoundException, kms.ErrCodeMalformedPolicyDocumentException)
 
 	if err != nil {
 		return fmt.Errorf("error updating KMS Key (%s) policy: %w", keyID, err)

--- a/internal/service/kms/key_test.go
+++ b/internal/service/kms/key_test.go
@@ -267,12 +267,6 @@ func TestAccKMSKey_Policy_iamRoleUpdate(t *testing.T) {
 					testAccCheckKeyExists(resourceName, &key),
 				),
 			},
-			//{
-			//	ResourceName:            resourceName,
-			//	ImportState:             true,
-			//	ImportStateVerify:       true,
-			//	ImportStateVerifyIgnore: []string{"deletion_window_in_days", "bypass_policy_lockout_safety_check"},
-			//},
 		},
 	})
 }

--- a/internal/service/kms/key_test.go
+++ b/internal/service/kms/key_test.go
@@ -244,6 +244,39 @@ func TestAccKMSKey_Policy_iamRole(t *testing.T) {
 	})
 }
 
+func TestAccKMSKey_Policy_iamRoleUpdate(t *testing.T) {
+	var key kms.KeyMetadata
+	rName := sdkacctest.RandomWithPrefix(acctest.ResourcePrefix)
+	resourceName := "aws_kms_key.test"
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:          func() { acctest.PreCheck(t) },
+		ErrorCheck:        acctest.ErrorCheck(t, kms.EndpointsID),
+		ProviderFactories: acctest.ProviderFactories,
+		CheckDestroy:      testAccCheckKeyDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccKey_policy(rName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckKeyExists(resourceName, &key),
+				),
+			},
+			{
+				Config: testAccKeyPolicyIAMRoleConfig(rName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckKeyExists(resourceName, &key),
+				),
+			},
+			//{
+			//	ResourceName:            resourceName,
+			//	ImportState:             true,
+			//	ImportStateVerify:       true,
+			//	ImportStateVerifyIgnore: []string{"deletion_window_in_days", "bypass_policy_lockout_safety_check"},
+			//},
+		},
+	})
+}
+
 // Reference: https://github.com/hashicorp/terraform-provider-aws/issues/11801
 func TestAccKMSKey_Policy_iamRoleOrder(t *testing.T) {
 	var key kms.KeyMetadata

